File: handiedge_engine/app/domain/ai_review/agents/shared.py
"""Shared building blocks for the three review agents.

Each agent follows the same shape:
  1. a deterministic rule pass (guardrails that always run), and
  2. an optional LLM reasoning pass (qualitative judgment).
This module holds the pieces both passes need: severity -> cap policy, verdict
assembly, and context serialization.
"""

from __future__ import annotations

import json

from app.domain.ai_review.confidence import min_rank
from app.domain.ai_review.provider import LlmVerdict
from app.domain.ai_review.types import (
    AgentRole,
    ReviewContext,
    ReviewFlag,
    ReviewRank,
    Severity,
)

_SEVERITY_WEIGHT: dict[Severity, int] = {
    Severity.INFO: 0,
    Severity.WARNING: 1,
    Severity.CRITICAL: 2,
}


def worst_severity(flags: list[ReviewFlag]) -> Severity | None:
    worst: Severity | None = None
    for flag in flags:
        if worst is None or _SEVERITY_WEIGHT[flag.severity] > _SEVERITY_WEIGHT[worst]:
            worst = flag.severity
    return worst


def cap_for_severity(severity: Severity | None) -> ReviewRank | None:
    """Default cap policy shared by the deterministic passes.

    A critical issue caps the pick at C (informational only); a warning caps it
    at B (no open-warning pick may be a headline S/A). Info-only findings impose
    no cap. Agents may override per-flag when a finding warrants a different one.
    """

    if severity is Severity.CRITICAL:
        return ReviewRank.C
    if severity is Severity.WARNING:
        return ReviewRank.B
    return None


def merge_caps(a: ReviewRank | None, b: ReviewRank | None) -> ReviewRank | None:
    """Combine two optional caps into the more conservative one."""

    if a is None:
        return b
    if b is None:
        return a
    return min_rank(a, b)


def llm_concerns_to_flags(agent: AgentRole, verdict: LlmVerdict) -> list[ReviewFlag]:
    """Tag each bare LLM concern with its owning agent to form a ReviewFlag."""

    return [
        ReviewFlag(agent=agent, severity=c.severity, code=c.code, message=c.message)
        for c in verdict.concerns
    ]


def serialize_context(ctx: ReviewContext) -> str:
    """Serialize a review context into a compact, faithful text block for the
    model. Keys are emitted in a fixed order so structurally identical games
    serialize consistently. The full view is included — reviewers decide what is
    relevant, we do not pre-filter."""

    def _starter(s) -> dict | None:
        return None if s is None else {"name": s.name, "confirmed": s.confirmed}

    view = {
        "matchId": ctx.match_id,
        "matchup": f"{ctx.away} @ {ctx.home}",
        "preReviewRank": ctx.original_tier_rank.value,
        "decisionIsPredict": ctx.is_predict,
        "data": {
            "scheduleConfirmed": ctx.schedule_confirmed,
            "homeStarter": _starter(ctx.home_starter),
            "awayStarter": _starter(ctx.away_starter),
            "battingStatsAvailable": ctx.batting_stats_available,
            "bullpenStatsAvailable": ctx.bullpen_stats_available,
            "recentFormAvailable": ctx.recent_form_available,
            "parkFactorsAvailable": ctx.park_factors_available,
            "oddsAvailable": ctx.odds_available,
            "weather": (
                None
                if ctx.weather is None
                else {
                    "windMph": ctx.weather.wind_mph,
                    "windDir": ctx.weather.wind_dir,
                    "tempF": ctx.weather.temp_f,
                }
            ),
            "injuries": [
                {
                    "player": i.player,
                    "side": i.side,
                    "status": i.status,
                    "keyPlayer": i.key_player,
                }
                for i in ctx.injuries
            ],
            "stalenessMinutes": ctx.staleness_minutes,
        },
        "model": {
            "homeWinProb": float(ctx.home_win_prob),
            "awayWinProb": float(ctx.away_win_prob),
            "componentAgreement": (
                None if ctx.component_agreement is None else float(ctx.component_agreement)
            ),
            "marketEdge": None if ctx.market_edge is None else float(ctx.market_edge),
            "predictedTotal": (
                None if ctx.predicted_total is None else float(ctx.predicted_total)
            ),
            "totalLine": None if ctx.total_line is None else float(ctx.total_line),
        },
        "keyFactors": list(ctx.key_factors),
    }
    return json.dumps(view, indent=2, ensure_ascii=False)

File: lib/ai-review/src/agents/shared.ts
/**
 * Shared building blocks for the three review agents.
 *
 * Each agent follows the same shape:
 *   1. a deterministic rule pass (guardrails that always run), and
 *   2. an optional LLM reasoning pass (qualitative judgment).
 * This module holds the pieces both passes need: context serialization,
 * severity→cap policy, and verdict assembly.
 */

import type {
  AgentRole,
  ConfidenceRank,
  GamePrediction,
  ReviewFlag,
  Severity,
} from "../types.js";
import type { LlmVerdict } from "../schema.js";
import { minRank } from "../confidence.js";

/** Numeric weight so we can pick the "worst" severity in a list. */
const SEVERITY_WEIGHT: Record<Severity, number> = {
  info: 0,
  warning: 1,
  critical: 2,
};

export function worstSeverity(flags: readonly ReviewFlag[]): Severity | null {
  let worst: Severity | null = null;
  for (const flag of flags) {
    if (worst === null || SEVERITY_WEIGHT[flag.severity] > SEVERITY_WEIGHT[worst]) {
      worst = flag.severity;
    }
  }
  return worst;
}

/**
 * Default cap policy shared by the deterministic passes:
 *  - a critical issue caps the pick at C (informational only), and
 *  - a warning caps it at B (can't be a headline S/A pick with an open warning).
 * Info-only findings impose no cap. Agents may override per-flag when a finding
 * warrants a different ceiling.
 */
export function capForSeverity(severity: Severity | null): ConfidenceRank | null {
  switch (severity) {
    case "critical":
      return "C";
    case "warning":
      return "B";
    default:
      return null;
  }
}

/** Combine two optional caps into the more conservative one. */
export function mergeCaps(
  a: ConfidenceRank | null,
  b: ConfidenceRank | null,
): ConfidenceRank | null {
  if (a === null) return b;
  if (b === null) return a;
  return minRank(a, b);
}

/** Tag a bare LLM concern with its owning agent to form a {@link ReviewFlag}. */
export function llmConcernsToFlags(
  agent: AgentRole,
  verdict: LlmVerdict,
): ReviewFlag[] {
  return verdict.concerns.map((c) => ({
    agent,
    severity: c.severity,
    code: c.code,
    message: c.message,
  }));
}

/** Normalize the LLM's `"none"` sentinel to a real cap (or null). */
export function llmRankToCap(
  suggested: LlmVerdict["suggestedMaxRank"],
): ConfidenceRank | null {
  return suggested === "none" ? null : suggested;
}

/**
 * Serialize a prediction into a compact, faithful text block for the model.
 * Keys are emitted in a fixed order so repeated reviews of structurally
 * identical games hash consistently. The full object is included — reviewers
 * decide what is relevant, we don't pre-filter.
 */
export function serializePrediction(pred: GamePrediction): string {
  const view = {
    gameId: pred.gameId,
    matchup: `${pred.away.abbreviation} @ ${pred.home.abbreviation}`,
    startTimeLocal: pred.startTimeLocal,
    preReviewConfidence: pred.confidence,
    data: pred.data,
    model: pred.model,
    keyFactors: pred.keyFactors ?? [],
  };
  return JSON.stringify(view, null, 2);
}

/** Round a fraction to a percentage string, e.g. 0.612 → "61.2%". */
export function pct(fraction: number): string {
  return `${(fraction * 100).toFixed(1)}%`;
}
